### ISSUE TITLE
Open files with O_EXCL in Python 2

### ISSUE THREAD

**Issue by sqwishy** [PyFilesystem/pyfilesystem2#182]
`io.open` supports taking file descriptor instead of a file path (and can take ownership of the fd). `os.open` lets you open a file in exclusive mode and returns a file descriptor. You could do something like the following to open the file in the mode specified.

```
diff --git a/fs/osfs.py b/fs/osfs.py
index e53e5e2..9b2c330 100644
--- a/fs/osfs.py
+++ b/fs/osfs.py
@@ -422,8 +422,8 @@ class OSFS(FS):
         _path = self.validatepath(path)
         sys_path = self._to_sys_path(_path)
         with convert_os_errors('open', path):
-            if six.PY2 and _mode.exclusive and self.exists(path):
-                raise FileExists(path)
+            if six.PY2 and _mode.exclusive:
+                sys_path = os.open(sys_path, os.O_CREAT | os.O_EXCL)
             _encoding = encoding or 'utf-8'
             return io.open(
                 sys_path,
```

*(I tried running the tests to see what this breaks, but for some reason tox wouldn't stop barfing at me even on regular v2.0.24 without this change.)*

**Comment by willmcgugan** [comment]
To avoid that extra call to `exists`? Seems like a good idea.

What was tox complaining about?

**Comment by sqwishy** [comment]
It's mainly since exclusive mode is supposed to be atomic. Using `exists()` allows a race condition that O_EXCL shouldn't have.

>What was tox complaining about?

Literally everything under the moon. Maybe my PATH is messed up or my version of gcc is older than time itself; I'm sure it's my environment. I can look more closely later today if it would help.

**Comment by sqwishy** [comment]
On a different machine, tox runs correctly, as far as I can tell, for 2.7 and 3.6 with four test failures in each version (`tests.test_copy.TestCopy`). It doesn't fail more when I apply the diff above.

The diff assumes that if you're in exclusive mode, you want to create. Which is a bit weird to me at first, but `open(2)` states
>In general, the behavior of O_EXCL is undefined if it is used without O_CREAT.  There is one exception: on Linux 2.6 and later, O_EXCL can be used without O_CREAT if pathname refers to a block device.  If the block device is in use by the system (e.g., mounted), open() fails with the error EBUSY.

So I think if you're using exclusive mode, then you're creating. Because you aren't using this library with paths to block devices. And also it's consistent with Python 3 where, if you `open()` with mode `x` and something else, instead of just `x` by itself, it will say `ValueError: must have exactly one of create/read/write/append mode`.

**Comment by sqwishy** [comment]
@willmcgugan Is there I can do to help this get merged in?

**Comment by willmcgugan** [comment]
A PR would be great...

**Comment by willmcgugan** [comment]
Fixed in 2.0.26
